### PR TITLE
Add 3D-file-viewer component

### PR DESCRIPTION
**Pull request by btzr-io**
### Changes
A basic ThreeJS 3D viewer  based on [Autodesk/3DViewerComponent
](https://github.com/Autodesk/3DViewerComponent)

- Add 3D file-viewing to LBRY desktop app https://github.com/lbryio/lbry-app/issues/1409

### Todo (WIP)
- [x] Threejs + React.
- [x] Load blob from claim
- [x] Show Grid
- [x] Render 3D-model.
- [x] Add 3D-file support for getMediaType function https://github.com/lbryio/lbry-redux/pull/41
- [X] Investigate render bug. https://github.com/mrdoob/three.js/issues/12499#issuecomment-339780204
- [x] Ensure that STL files have valid normal data.
- [x]  Show loading  message
- [x] Handle Errors
- [x] Adaptive Theme (Dark / Light) :art: 
- [x] STL loader support.
- [ ] OBJ loader support.

### CODE REVIEW

**Comment by seanyesmunt** [comment]
@btzr-io Is this ready for review?

**Comment by btzr-io** [comment]
@seanyesmunt I think we should use a progress bar instead of the current spinner for the loading screen:
What do you think?

**Comment by btzr-io** [comment]
@seanyesmunt So I think it's ready for review, 
I just need to add some basic controls (UI) and support more file types (currently STL only).

**Comment by seanyesmunt** [comment]
Will check this out today!

**Comment by btzr-io** [comment]
ok, I'll setup a test branch...

**Comment by btzr-io** [comment]
Test branch: https://github.com/lbryio/lbry-app/tree/test-3d

**Comment by seanyesmunt** [comment]
Hey @btzr-io. Probably won't get a chance to look at this until the redesign is live. This will be one of the first things that go in after that though.

**Comment by seanyesmunt** [comment]
Ok! Ready to get this in. I'm seeing `Sorry, looks like we can't play this file.` when I go to the file in your description. Not sure if this is from a recent change or your code.

Is this dependent on https://github.com/lbryio/lbry-app/pull/1576?

**Comment by seanyesmunt** [comment]
Oh I see that PR references this component, so it's the other way around. Ignore me

**Comment by btzr-io** [comment]
@seanyesmunt where did you test it? check the test-branch: https://github.com/lbryio/lbry-app/pull/1558#issuecomment-396812301

**Comment by tzarebczan** [comment]
@btzr-io can you try lbry://stl? Not seeing anything but the grid when it first opens. Went out and back into the claim, then it worked.

**Comment by btzr-io** [comment]
@tzarebczan fixed in https://github.com/lbryio/lbry-app/pull/1576/commits/c769a47babf6df3ea41e45ffff6ad64094f72f8e

**Comment by tzarebczan** [comment]
@btzr-io tried an stl file yesterday on this branch and got "this file cannot be played".

**Comment by btzr-io** [comment]
@tzarebczan There is no implementation of the component yet :upside_down_face:

**Comment by seanyesmunt** [comment]
Planning to merge this sometime this soon.

File: src/renderer/component/threeViewer/index.jsx [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+  startLoader() {
+    const { source } = this.props;
+
+    if (source) {
+      ThreeLoader(source, this.renderModel.bind(this), {
+        onStart: this.handleStart(this),
+        onLoad: this.handleReady.bind(this),
+        onError: this.handleError.bind(this),
+        onProgress: this.handleProgress.bind(this),
```
Comment (seanyesmunt) [inline src/renderer/component/threeViewer/index.jsx]: You shouldn't need to bind `this` in this functions. `handleStart` (and the others) uses the same `this` reference

File: src/renderer/component/threeViewer/index.jsx [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    this.renderer.setSize(width, height);
+  };
+
+  handleError() {
+    this.setState({ error: "Sorry, looks like we can't load this file" });
+  }
+
+  handleProgress() {
+    // const progress = (currentItem / totalItems) * 100;
+    // console.info(currentItem, totalItems, progress);
```
Comment (seanyesmunt) [inline src/renderer/component/threeViewer/index.jsx]: I think we will hold off on the progress bar for now, we have a progress percentage below the file. 

I am going to add analytics for start to stream time, that will give us an idea of average time spent looking at the loading screen. If it's decently high I think we can add it.

File: src/renderer/component/threeViewer/internal/loader.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+const Manager = ({ onLoad, onStart, onProgress, onError }) => {
+  const manager = new LoadingManager();
+  manager.onLoad = onLoad;
+  manager.onStart = onStart;
+  manager.onProgress = onProgress;
+  manager.onError = onError;
+
+  return manager;
+};
```
Comment (seanyesmunt) [inline src/renderer/component/threeViewer/internal/loader.js]: I don't think this function is being used? I see it being called below, but that value isn't being used.

File: src/renderer/component/fileRender/view.jsx [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import React from 'react';
 import LoadingScreen from 'component/common/loading-screen';
 import PdfViewer from 'component/viewers/pdfViewer';
+import ThreeViewer from 'component/threeViewer';
```
Comment (seanyesmunt) [inline src/renderer/component/fileRender/view.jsx]: I added this so it works, this component should probably live inside of `component/viewers/...` to keep it consistent.

**Comment by seanyesmunt** [comment]
We are seeing some issues where the app becomes super un-responsive after viewing 3d files (I've had to restart twice after my computer freezes).

Will continue to test, but it should be simple stuff to help improve performance.

**Comment by skhameneh** [comment]
Hey @btzr-io can you add debug info for geometry count?
@seanyesmunt is having trouble when working with a demo, his machine freezes when leaving the viewer.

I'd also like to note that it might be good to explore simplifying complex geometries before rendering:
https://threejs.org/examples/?q=simplifi#webgl_modifier_simplifier

**Comment by seanyesmunt** [comment]
Possibly realated to running in dev mode. I shared the binaries internally and didn't hear any issues. I'll try running those.

**Comment by seanyesmunt** [comment]
Seems to work fine with a binary. I'm guessing it was just from running in dev mode. Maybe we can simplify the geometry in dev mode? That should help.

File: src/renderer/component/threeViewer/internal/loader.js (hunk 3 above)
Comment (btzr-io) [reply src/renderer/component/threeViewer/internal/loader.js]: https://github.com/lbryio/lbry-desktop/blob/34ba2cf3ebad8383e045d606ff8b5c27921e53a7/src/renderer/component/threeViewer/internal/loader.js#L24

**Comment by btzr-io** [comment]
>We are seeing some issues where the app becomes super un-responsive after viewing 3d files (I've had to restart twice after my computer freezes).

Not sure what you mean by after, the only time I'm seeing this issue is while loading the model:
> Loading 3D model...

Looks like the loader is blocking the UI until the loading and parsing process of the file is completed 
https://stackoverflow.com/questions/48636384/web-workers-for-loading-object-in-three-js


>  Maybe we can simplify the geometry in dev mode?

I tried to follow the `webgl_modifier_simplifier` example but the model disappears from the scene.

**Pull request merged**